File: eval/rule_accuracy/corpus/python/CWE-89/tn_safe_queries.py
# SQL injection — true negatives (safe patterns)
# no-expect: These should NOT trigger any SQL injection rules

import sqlite3

def safe_parameterized(user_id):
    cursor = sqlite3.connect(":memory:").cursor()
    cursor.execute("SELECT * FROM users WHERE id = ?", (user_id,))

def safe_named_params(name, email):
    cursor = sqlite3.connect(":memory:").cursor()
    cursor.execute(
        "SELECT * FROM users WHERE name = :name AND email = :email",
        {"name": name, "email": email},
    )

def safe_constant():
    cursor = sqlite3.connect(":memory:").cursor()
    cursor.execute("SELECT count(*) FROM users")

def safe_orm_style(user_id):
    # ORM-style — no raw SQL
    return User.objects.filter(id=user_id).first()

File: eval/rule_accuracy/corpus/go/CWE-89/tn_safe_queries.go
// SQL — true negatives (safe patterns)
// no-expect: These should NOT trigger SQL injection rules
package corpus

import "database/sql"

func safeParameterized(db *sql.DB, userID string) {
	db.Query("SELECT * FROM users WHERE id = ?", userID)
}

func safePrepared(db *sql.DB, name string) {
	stmt, _ := db.Prepare("SELECT * FROM users WHERE name = ?")
	stmt.Query(name)
}

func safeConstant(db *sql.DB) {
	db.Query("SELECT count(*) FROM users")
}
